Make package updates optional

codificat opened the pull request:
Add an oo_option "skip_pkg_update" that allows to skip package updates (i.e. do not "yum update *" if this is set to something different than false)

This is useful when provisioning reproducer environments that should keep certain package versions

menren commented:
Nice option +1

jaryn commented:
Can one of the admins verify this patch?

sdodson commented:
That turns this role into a NOOP, I think this logic should be pushed outside of this role perhaps into the two playbooks that call this role.

playbooks/byo/rhel_subscribe.yml
playbooks/common/openshift-cluster/update_repos_and_packages.yml

jaryn commented:
Can one of the admins verify this patch?

sdodson commented:
@codificat bump

arilivigni commented:
Testing job has been triggered.

arilivigni commented:
Internal Jenkins error during triggering.
 No test results found.

codificat commented:
@sdodson I'm sorry for the delay here. Moved the logic to the two playbooks as suggested.

I was wondering though: aren't these two playbooks too similar to each other? I.e. shouldn't we somehow do the "move" the other way around to reduce redundancy, i.e. consolidate them? I guess though that the playbook inside byo can be used "standalone"...

detiber commented:
@codificat There does appear to be quite a bit of overlap between the two playbooks.

In general the playbooks under playbooks/common should be sufficiently generic so that they are not tied to a specific type of inventory/provider specific playbook.

The playbooks in the provider specific directories (playbooks/{aws,byo,gce,openstack,...}) should provide a minimal wrapper to include the common playbook(s) and setting the required variables that may differ based on provider.  https://github.com/openshift/openshift-ansible/blob/master/playbooks/byo/openshift-cluster/config.yml as an example.

Completely untested, but I would think this should work for playbooks/byo/rhel_subscribe.yml:
```
---
- name: Populate config host groups
  hosts: localhost
  gather_facts: no
  tasks:
  - name: Evaluate oo_hosts_to_update
    add_host:
      name: "{{ item }}"
      groups: oo_hosts_to_update
      openshift_cluster_id: "{{ cluster_id | default('default') }}"
      openshift_debug_level: 2
      openshift_deployment_type: "{{ deployment_type }}"
    with_items: groups.masters | union(groups.nodes | union(groups.etcd | union(groups.lb)))

- include: ../../common/openshift-cluster/update_repos_and_packages.yml
```

codificat commented:
Trying to finally complete this one... thanks for the comments @detiber, I rebased and updated `rhel_subscribe.yml` basically as you suggested but using `openshift-cluster/cluster_hosts.yml` that is now available.
Therefore this version adds the `skip_pkg_update` option to the `update_repos_and_packages.yml` playbook and removes the overlap with rhel_subscribe.

detiber commented on playbooks/common/openshift-cluster/update_repos_and_packages.yml:
```diff
@@ -9,4 +9,6 @@
   - openshift_repos
-  - os_update_latest
+  - role: os_update_latest
+    when: lookup('oo_option', 'skip_pkg_update') | default('no', True) |
```
This should probably be something like:
```
when: openshift_skip_pkg_update | default(lookup('oo_option', 'skip_pkg_update') | default(False, True) | bool
```

That way users using the byo playbooks can use openshift_skip_pkg_update instead of cli_skip_pkg_update while retaining support for the bin/cluster options and support all of the true/false values supported by ansilbe (bool filter).

detiber commented:
@codificat thanks! Outside of the one comment I have around the conditional used for applying the os_update_latest role this is looking great.

@simon3z This is relevant to your interests.

codificat commented:
Thanks for the feedback @detiber , updated the conditional used for the `os_update_latest` role. Besides your comment I also excluded it during a containerized install (you can't update that).

Also applied a similar modification to `rhel_skip_subscribtion` in the same file, and added both options as comments to the sample aep and ose BYO inventories.

detiber commented on inventory/byo/hosts.aep.example:
```diff
@@ -24,6 +24,14 @@ debug_level=2
+# By default the installer attempts to update the system before installing
+# and configuring. Set openshift_skip_pkg_update=true to prevent that update
+#openshift_skip_pkg_update=false
+
```
I'm not against documenting these here, however it should probably also mention that the playbooks/common/openshift-cluster/update_repos_and_packages.yml isn't called by default when running playbooks/byo/config.yml.

detiber replied:
There should also be some documentation in the playbook of the variables to set for host subscription as well.

detiber commented on playbooks/byo/rhel_subscribe.yml:
```diff
@@ -1,12 +1,19 @@
-          default('no', True) | lower in ['no', 'false']
-  - openshift_repos
-  - os_update_latest
+- name: Populate oo_hosts_to_update group
```
This playbook should probably be moved under playbooks/byo/openshift-cluster.

detiber commented on playbooks/common/openshift-cluster/update_repos_and_packages.yml:
```diff
@@ -5,8 +5,10 @@
-            default('no', True) | lower in ['no', 'false']
+          ansible_distribution == "RedHat" and not ( openshift_rhel_skip_subscription |
+          default(lookup('oo_option', 'rhel_skip_subscription')) | default(rhsub_skip) |
+          default(False) | bool )
```
might make sense to break some of this out into the vars section to clean up this when clause a bit.

Also, the value of rhsub_skip and the default(False) will never be used here.

Warning, the below is completely untested:
```
vars:
...
  # store this value to conditionally set it only if it doesn't return '', since the oo_option
  # lookup filter will return '' if the variable isn't set.
  t_oo_opt_skip_sub: "{{ lookup('oo_option', 'rhel_skip_subscription') }}"
  # to maintain the ordering you indicated above, we need to work around the fact that
  # the oo_option lookup plugin will return '' if the env var/variable is not defined.
  t_skip_sub_val: "{{ openshift_rhel_skip_subscription
                                | default(t_oo_opt_skip_sub if t_oo_opt_skip_sub != '' else rhsub_skip | default(False)) }}"
  t_valid_dt_for_sub: "{{ deployment_type in ['enterprise', 'atomic-enterprise', 'openshift-enterprise'] }}"
  t_skip_rhel_subscribe: "{{ t_valid_dt_for_sub | bool and ansible_distribution == 'RedHat' and not t_skip_sub_val | bool }}"
```

It can be cleaned up a bit further, by using a conditional set_fact task to define openshift_rhel_skip_subscription as the value of the oo_option lookup on rhel_subscription if it isn't already defined.
```
vars:
...
  t_oo_opt_rhel_skip_sub: "{{ lookup('oo_option', 'rhel_skip_subscription') }}"
  t_valid_dt_for_sub: "{{ deployment_type in ['enterprise', 'atomic-enterprise', 'openshift-enterprise'] }}"
...
pre_tasks:
- set_fact:
    openshift_rhel_skip_subscription: "{{ t_oo_opt_rhel_skip_sub if t_oo_opt_rhel_skip_sub != '' else rhsub_skip | default(False) }}"
  when: openshift_rhel_skip_subscription is not defined
...
roles:
- role: rhel_subscribe:
  when: t_valid_dt_for_sub | bool and ansible_distribution == 'RedHat' and not openshift_rhel_skip_subscription | bool
```

detiber replied:
Also, if using the set fact to merge the oo_option lookup vars under the openshift_* vars, you can push the conditional logic down into the role itself instead of conditionalizing the role inclusion in the playbook (since the when on the role simply applies the when conditional to all tasks in the role anyway).

detiber commented:
@codificat my apologies about all of the feedback on this, it just happens that we are looking at leveraging this work with the upcoming ManageIQ integration.

Some other items that will need to be done for that integration (most definitely not blockers for this PR, just jotting them down while they are on my mind).
- Support for using an activation key for registration/subscription
- Support for configuring the subscriptions based on one or more pool-ids given
- Support for configuring stack-able subscriptions
- Support for specifying/overriding the list of repos to enable/disable to support cloned repos, enabling features that may require additional subscriptions, handling a mixed use case of RHUI and RHSM for repos.

detiber commented:
Closing due to age.
